scripts/prepare_doc_publication.py

```diff
@@ -0,0 +1,136 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Prepares MarkDown documentation publication.
+
+The in-tree and auto-generated MarkDown documentation lacks necessary metadata
+(i.e., front matter) for specifying the layout, title, and others that are
+required by Jekyll for rendering the HTML. This script patches MarkDown
+files with that.
+"""
+
+import argparse
+import os
+
+
+def parse_arguments():
+    """Parses command-line options."""
+    parser = argparse.ArgumentParser(description='Processes MarkDown files for publication')
+    parser.add_argument('base_dir', metavar='PATH', type=str, help='Base documentation directory.')
+
+    args = parser.parse_args()
+    if not os.path.isdir(args.base_dir):
+        raise parser.error('expected path to a directory')
+
+    return args
+
+# A dictionary containing source file to permanent link mappings.
+#
+# By default a source file will have a permanent URL link following its
+# filename. For example, if we have docs/Foo/Bar.md, then the permanent link
+# for it would be https://google.github.io/iree/Foo/Bar. This dictionary
+# allows one to override the permanent link if necessary.
+PERMALINK_DICT = {
+    'roadmap_design.md': "DesignRoadmap",
+}
+
+# A dictionary containing source file to navigation order mappings.
+#
+# By default the rendered docs will be sort alphabetically and listed on
+# the left panel of https://google.github.io/iree website. This allows one
+# to specify an order for a specific doc.
+NAVI_ORDER_DICT = {
+    # 'Home' is 1.
+    'roadmap_design.md': 2,
+}
+
+def process_file(basedir, relpath, filename):
+    """Modifies the given file in-place with metadata and other changes to
+    make it suitable for publication."""
+
+    full_path = os.path.join(basedir, relpath, filename)
+    base_name = os.path.splitext(filename)[0]
+    with open(full_path, 'r') as f:
+        content = f.read()
+
+    # Directly return if the file already has front matter.
+    if content.startswith('---\n'):
+        return
+
+    front_matter = {}
+    # Use the default layout for everything.
+    front_matter['layout'] = 'default'
+    # Use the base filename as permanent link.
+    front_matter['permalink'] = base_name
+
+    # Organize each doc to a section matching its directory structure.
+    if relpath and relpath != '.':
+        front_matter['parent'] = relpath
+        front_matter['permalink'] = '{}/{}'.format(relpath, front_matter['permalink'])
+
+    # Find the title and TOC.
+    lines = content.splitlines()
+    title_line_index = None
+    toc_index = None
+    for (index, line) in enumerate(lines):
+        if line.startswith('# '):
+            title_line_index = index
+        if line == '[TOC]':
+            toc_index = index
+        if title_line_index is not None and toc_index is not None:
+            break
+
+    # Replace '[TOC]' with the proper format that can be rendered.
```

Comment:
I don't see this already in the repo. Is this what's generated by the script. Could we have it generate the proper thing instead? I think internally we use this format, but if we want to get this working internally we can use copybara to rewrite the generator script.

Reply:
This is generated by `mlir-tblgen`'s DocGen backend. I think this is what accepted by mlir.llvm.org. 

Do we want to publish the generated docs also internally? I don't really see the benefits of doing that. having https://google.github.io/iree suffices.

Reply:
I was just saying that "[TOC]" is supported internally and wasn't sure if that's why you were using it here. I see this is generated by a tool we don't control.

Reply:
Digged a bit into this. mlir-www also substitutes this with something their website generator (Hugo) supports: https://github.com/llvm/mlir-www/blob/master/copy_docs.sh#L36. So we are not so different here. :)